sklearn/calibration.py

```diff
@@ -870,6 +880,71 @@ def predict(self, T):
         return expit(-(self.a_ * T + self.b_))
 
 
+class _CustomCalibration(RegressorMixin, BaseEstimator):
+    """Calibration using a custom classifier.
+
+    Parameters:
+    -----------
+    method : object with 'fit' and 'predict_proba' methods
+        The classifier to use for the calibration.
+    """
+
+    def __init__(self, method):
+        self.method = clone(method)
+
+    def fit(self, X, y, sample_weight=None):
+        """Fit the model using X, y as training data.
+
+        Parameters
+        ----------
+        X : array-like of shape (n_samples,) or (n_samples, 1)
+            Training data.
+
+        y : array-like of shape (n_samples,)
+            Training target.
+
+        sample_weight : array-like of shape (n_samples,), default=None
+            Weights. If set to None, all weights will be set to 1 (equal
+            weights).
+
+        Returns
+        -------
+        self : object
+            Returns an instance of self.
+        """
+        X = X.reshape(-1, 1)
+        sample_weight = _check_sample_weight(sample_weight, X, dtype=X.dtype)
+
+        # Discard samples with null weights
+        mask = sample_weight > 0
+        X, y, sample_weight = X[mask], y[mask], sample_weight[mask]
+
+        self.method.fit(X, y, sample_weight)
+        return self
+
+    def predict(self, T):
+        """Predict calibrated probabilities using the 'predict_proba' method
+        of the classifier.
+
+        Parameters
+        ----------
+        T : array-like of shape (n_samples,) or (n_samples, 1)
+            Data to transform.
+
+        Returns
+        -------
+        y_pred : ndarray of shape (n_samples,)
+            Transformed data.
+        """
+        T = T.reshape(-1, 1)
+        probas = self.method.predict_proba(T)
+
+        # If binary classification, only return proba of the positive class
+        if probas.shape[1] == 2:
+            return probas[:, 1]
+        return probas
```

Comment:
We need a new test to cover for the multiclass case, for instance using `LogisticRegression(C=1e12)`.